### ISSUE TITLE
Errors on Debian in docker

### ISSUE THREAD

**Issue by romkatv** [poetaman/arttime#41]
1. Run this:
    ```zsh
    docker run -e TERM -e LC_ALL=C.UTF-8 -it --rm zshusers/zsh:5.8 zsh -c '
      apt update
      apt install -y curl
      curl -fsSL https://gist.githubusercontent.com/poetaman/bdc598ee607e9767fe33da50e993c650/raw/d3d5e29427304d977044b594fbb6cc84616c30d4/arttime_online_installer.sh | zsh
      typeset -p TERM
      exec ~/.local/bin/arttime'
    ```
2. Press <kbd>q</kbd>.
3. Press <kbd>y</kbd>.

Expected: no errors.

Actual:

```text    
DEPENDS: If you want desktop notifications, you need notify-send.
DEPENDS: If you want desktop sounds, you need one of: 1) pulseaudio daemon
         running, or 2) pipewire daemon running, or 3) vorbis-tools.
basename: missing operand
Try 'basename --help' for more information.
Installation *almost* complete! To start using arttime, follow these steps:
    1) Add /root/.local/bin to your PATH environment variable in appropriate file,
    2) Open a new terminal session, type 'arttime' and press Enter.
To run it right away from this shell, execute arttime by specifying its full path:
       /root/.local/bin/arttime
export TERM=screen-256color
/root/.local/bin/arttime:echoti:500: no such terminfo capability: rmam
/root/.local/bin/arttime:echoti:502: no such terminfo capability: smam
trapexit_blocking:1: command not found: pkill
```

**Comment by poetaman** [comment]
@romkatv Thanks for reporting! I never tested this on docker. I have tested this on Debian and many other Linux/BSD VMs, and have never seen these errors with commands ``basename`` or ``pkill``. The ``terminfo`` capability errors are understandable as base installation has arcane terminfos (and misrepresents the capability that your terminal actually has: wrapping). ``pkill`` and ``basename`` come preinstalled under ``/usr/bin/``. What is the output of ``uname -a`` on your docker's Debian? My Debian VM's ``uname -a`` output is below. A first suspicion is that there is a lacking docker configuration, it's a little late here or I would have dug a bit more immediately...

```
parallels@debian-gnu-linux-11:~$ uname -a
Linux debian-gnu-linux-11 5.10.0-15-arm64 #1 SMP Debian 5.10.120-1 (2022-06-09) aarch64 GNU/Linux
```

**Comment by romkatv** [comment]
Here's basename:

```text
% which basename
/usr/bin/basename

% basename --version
basename (GNU coreutils) 8.30
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by David MacKenzie.
```

Here's uname:

```text
% uname -a
Linux b7fe11a26cde 5.4.72-microsoft-standard-WSL2 #1 SMP Wed Oct 28 23:40:43 UTC 2020 x86_64 GNU/Linux
```

Here's terminfo:

```text
% infocmp
#       Reconstructed via infocmp from file: /lib/terminfo/s/screen-256color
screen-256color|GNU Screen with 256 colors,
        am, km, mir, msgr, xenl,
        colors#0x100, cols#80, it#8, lines#24, pairs#0x10000,
        acsc=++\,\,--..00``aaffgghhiijjkkllmmnnooppqqrrssttuuvvwwxxyyzz{{||}}~~,
        bel=^G, blink=\E[5m, bold=\E[1m, cbt=\E[Z, civis=\E[?25l,
        clear=\E[H\E[J, cnorm=\E[34h\E[?25h, cr=\r,
        csr=\E[%i%p1%d;%p2%dr, cub=\E[%p1%dD, cub1=^H,
        cud=\E[%p1%dB, cud1=\n, cuf=\E[%p1%dC, cuf1=\E[C,
        cup=\E[%i%p1%d;%p2%dH, cuu=\E[%p1%dA, cuu1=\EM,
        cvvis=\E[34l, dch=\E[%p1%dP, dch1=\E[P, dim=\E[2m,
        dl=\E[%p1%dM, dl1=\E[M, ed=\E[J, el=\E[K, el1=\E[1K,
        enacs=\E(B\E)0, flash=\Eg, home=\E[H, hpa=\E[%i%p1%dG,
        ht=^I, hts=\EH, ich=\E[%p1%d@, il=\E[%p1%dL, il1=\E[L,
        ind=\n, indn=\E[%p1%dS, is2=\E)0, kbs=^?, kcbt=\E[Z,
        kcub1=\EOD, kcud1=\EOB, kcuf1=\EOC, kcuu1=\EOA,
        kdch1=\E[3~, kend=\E[4~, kf1=\EOP, kf10=\E[21~,
        kf11=\E[23~, kf12=\E[24~, kf2=\EOQ, kf3=\EOR, kf4=\EOS,
        kf5=\E[15~, kf6=\E[17~, kf7=\E[18~, kf8=\E[19~, kf9=\E[20~,
        khome=\E[1~, kich1=\E[2~, kmous=\E[M, knp=\E[6~, kpp=\E[5~,
        nel=\EE, op=\E[39;49m, rc=\E8, rev=\E[7m, ri=\EM, rmacs=^O,
        rmcup=\E[?1049l, rmir=\E[4l, rmkx=\E[?1l\E>, rmso=\E[23m,
        rmul=\E[24m, rs2=\Ec\E[?1000l\E[?25h, sc=\E7,
        setab=\E[%?%p1%{8}%<%t4%p1%d%e%p1%{16}%<%t10%p1%{8}%-%d%e48;5;%p1%d%;m,
        setaf=\E[%?%p1%{8}%<%t3%p1%d%e%p1%{16}%<%t9%p1%{8}%-%d%e38;5;%p1%d%;m,
        sgr=\E[0%?%p6%t;1%;%?%p1%t;3%;%?%p2%t;4%;%?%p3%t;7%;%?%p4%t;5%;%?%p5%t;2%;m%?%p9%t\016%e\017%;,
        sgr0=\E[m\017, smacs=^N, smcup=\E[?1049h, smir=\E[4h,
        smkx=\E[?1h\E=, smso=\E[3m, smul=\E[4m, tbc=\E[3g,
        vpa=\E[%i%p1%dd,
```

All of these look fine to me, including terminfo. screen-256color simply doesn't have `rmam` or `smam`.

The advantage of my posting the docker command is that you don't have to ask these questions. The instructions to reproduce are complete and self-contained. You can try them on any OS where you can install docker.

`pkill` is not POSIX, so it's not guaranteed to exist. If `arttime` cannot function without it, it's a good idea to issue an error upfront. For example:

```text
% arttime
error: please install pkill and try again.
```

**Comment by poetaman** [comment]
@romkatv Ah I see, I have never used docker so I did not know it was self contained. Thanks for the explanation. You are right, pkill is not posix, though majority of systems come with it: All Linuxes, BSD Unixes, macOS. I see you are using Windows Subsystem for Linux, and seems like it doesn't come with pkill. Regarding screen-256color not having smam/rmam, users of modern terminal emulators should either create their custom screen-256color-real or simply not use screen-256color as modern terminal emulators do have wrapping capability. Terminfo database presumes that a user knows what capabilities their terminal has, and expects that the user maintain a capability database themselves. In post-real-terminal era, we users take terminfo for granted. If you don't feel motivated to edit and maintain your terminfo, then arttime with xterm-256color.

I fear I won't be able to run a WSL Debian environment to test on my Apple Silicon ARM Mac. I have Windows VM on Parallels, but it doesn't support WSL in it as Apple M1 chips don't support nested virtualization... Not sure if docker has ARM image for WSL in Debian.

**Comment by romkatv** [comment]
Feel free to close this issue. It won't upset me. I filed it because it's the right thing to do when when open source software prints unintended errors and I hope that others would do the same when using my own software. My workflow doesn't depend on your making any changes to arttime as a response to this issue.

A few tips should you wish to do something here:

- You can use use `${foo:t}` instead of `$(basename -- $foo)". It's a lot faster and avoids portability pitfalls.
- You can use use `(( ${+terminfo[rmam] ))` after `zmodload zsh/terinfo` to detect terminfo capabilities.
- You can use `(( ${+commands[pkill]} ))` to detect the availability of commands. It's faster than the common `command -v` because you don't need to redirect output.
- If arttime requires zsh >= 5.4, you can use `[[ -v foo ]]` instead of `$(( ${+foo} ))`.

I haven't looked at the source code of arttime, so I don't know how applicable these tips are.

**Comment by romkatv** [comment]
> I fear I won't be able to run a WSL Debian environment to test on my Apple Silicon ARM Mac.

Forgot to reply to this. You can install docker on macOS and then run the command I posted. It's not relevant that I'm using a different host OS, otherwise I would've specified it in the issue.

Note that my `TERM` is `screen-256color`. If you want to fully reproduce the issue, start `screen` or `tmux` on the host and then run the `docker` command. Your `TERM` may be `screen`, `screen-256color` or `tmux-256color` depending on your scree/tmux config. All of these will trigger the same errors in arttime.

**Comment by poetaman** [comment]
> Feel free to close this issue. It won't upset me. I filed it because it's the right thing to do when when open source software prints unintended errors and I hope that others would do the same when using my own software.

Let's keep this open, I appreciate your valuable technical feedback. Yes open source software thrives on such collaboration. This motivates me to try my hands on docker too.

Regarding zsh niceties, great observations. While starting to write arttime I wanted to code it in a way that I could easily port it to bash if I had wanted to. Over time I became biased to just keep using the more amazing features of zsh. Also I tend to do both bash and zsh, and if I am writing a snippet which I can directly apply to both, I tend to code it in a way that works for both (my industry still requires me to do bash). Though am always open to best syntactic sugar and features for performance critical parts of an application (as that's where user really sees value), and value such feedback.

Will get back sometime in coming days after I think I have resolved the main issues you faced. It would be good if you can give it a try again then.

**Comment by poetaman** [comment]
This is not an arttime issue. I was able to make arttime work under docker with changed settings/config. Update your docker command to following, and you will be able to try arttime.

```
docker run -e TERM -e LC_ALL=C.UTF-8 -e TZ -it --rm zshusers/zsh:5.8 zsh -c '
  export SHELL=$(command -v zsh)
  apt update
  apt install -y curl less procps
  curl -fsSL https://gist.githubusercontent.com/poetaman/bdc598ee607e9767fe33da50e993c650/raw/d3d5e29427304d977044b594fbb6cc84616c30d4/arttime_online_installer.sh | TERM=xterm-256color zsh
  TERM=xterm-256color ~/.local/bin/arttime;'
```

Here's my assessment:
* Docker containers are pretty bare-bones. For instance, even $SHELL environment variable is not set in it. It is set in GRUB after Debian boots in my VM. Setting $SHELL fixed basename command issue.
* If a user can do ``apt install -y curl``, then can get commands like ``pkill`` too. I updated the apt install command above to include ``procps``. It's not necessary for arttime to function, but I will anyway redirect error to null.
* Base Linux/BSD OS installations come with arcane terminfo/termcap databases. On my computer I have ``tmux-256color`` with ``smam``, ``rmam`` and it works. As mentioned in an earlier comment, we as modern terminal emulator users have been pampered to not care about maintaining our own terminfos that reflects true capabilities of our terminal emulator applications. Nevertheless, in above docker config I changed TERM to ``xterm-256color`` for arttime's commands for the users who want to quickly try arttime in docker. It works on my end from within tmux or outside.
* ``TZ`` is inherited too if set, if not arttime will start in UTC timezone. Docker doesn't seem to have a clean interface to inherit timezone info. Mount only partially seems to work.
